File: backend/chatbot.py
"""
Chatbot analítico con LLM gratuito (Groq primario, Gemini fallback).

Patrón RAG liviano: en cada llamada se inyecta un contexto JSON (métricas precalculadas
relevantes a la pregunta) dentro del system prompt. Stateless: el frontend manda el
historial completo en cada request. Si no hay API keys, devuelve un mensaje amigable
(nunca rompe con 500).
"""
from __future__ import annotations

import json
import os
import re

import httpx
from dotenv import load_dotenv

load_dotenv()

GROQ_URL = "https://api.groq.com/openai/v1/chat/completions"
GEMINI_URL = "https://generativelanguage.googleapis.com/v1beta/openai/chat/completions"

GROQ_MODEL = "llama-3.3-70b-versatile"
GEMINI_MODEL = "gemini-2.0-flash"


def build_system_prompt(contexto: dict) -> str:
    return f"""Eres un analista de operaciones de Arca Continental especializado en el
sistema Order Rescue. Tu rol es ayudar a supervisores de CEDIS, gerentes de operaciones
y equipos comerciales a entender los patrones de sustitución de pedidos.

Responde siempre en español, de forma clara y concisa. Cuando des números, sé preciso.
Si no tienes información suficiente para responder, dilo honestamente.

CONTEXTO ACTUAL DEL SISTEMA (datos reales):
{json.dumps(contexto, ensure_ascii=False, indent=2)}

GLOSARIO:
- CEDIS: Centro de Distribución
- Sustitución: cuando el producto pedido no está disponible y se reemplaza por otro
- Tasa de afectación: % de pedidos de un CEDIS que tuvieron al menos 1 sustitución
- Score de riesgo: probabilidad estimada de sustitución (0-1), verde <0.08, amarillo 0.08-0.14, rojo >0.14
- S1: señal por (CEDIS, SKU), S2: señal global por SKU, S3: señal por CEDIS

Sé útil, directo y orientado a acción. Cuando identifiques un problema, sugiere qué hacer."""


def get_context_for_query(query: str, app_state: dict) -> dict:
    """Router de contexto: según la pregunta decide qué datos inyectar."""
    q = query.lower()
    eda = app_state["eda"]
    contexto = {"metricas_globales": eda["metricas_globales"]}

    if any(w in q for w in ["cedis", "centro", "distribuc", "riesgo", "problema", "zona", "rojo"]):
        contexto["top_cedis_riesgosos"] = eda["top_cedis"][:10]

    if any(w in q for w in ["sku", "producto", "sustitu", "reemplaz", "coca", "topo", "ciel", "agua", "leche", "yogurt"]):
        contexto["top_skus_sustituidos"] = eda["top_skus_por_tasa"][:10]
        contexto["top_pares_sustitucion"] = eda["top_pares"][:10]

    if any(w in q for w in ["país", "pais", "méxico", "mexico", "ecuador", "perú", "peru", "argentina"]):
        contexto["por_pais"] = eda["por_pais"]

    # Detalle de pedido si menciona un ID largo
    match = re.search(r"\d{10,}", query.replace(".", "").replace(",", ""))
    if match:
        try:
            id_pedido = float(match.group())
            ps = app_state["scoring"].pedido_score.get(id_pedido)
            if ps:
                contexto["pedido_consultado"] = {"id_pedido": match.group(), **ps}
        except Exception:
            pass

    return contexto


async def call_llm(messages: list) -> tuple[str, str | None]:
    """
    Llama al LLM con fallback Groq -> Gemini.
    Devuelve (respuesta, proveedor_usado). proveedor_usado None si todo falló.
    """
    providers = [
        {"name": "Groq · llama-3.3-70b", "url": GROQ_URL, "model": GROQ_MODEL,
         "key": os.getenv("GROQ_API_KEY", "").strip()},
        {"name": "Gemini · 2.0-flash", "url": GEMINI_URL, "model": GEMINI_MODEL,
         "key": os.getenv("GEMINI_API_KEY", "").strip()},
    ]

    if not any(p["key"] for p in providers):
        return ("⚠️ El asistente de IA no está configurado todavía. Agrega tu GROQ_API_KEY "
                "o GEMINI_API_KEY al archivo .env y reinicia el servidor para activarlo. "
                "Mientras tanto, puedes usar los tableros del dashboard.", None)

    last_error = None
    async with httpx.AsyncClient(timeout=30.0) as client:
        for p in providers:
            if not p["key"]:
                continue
            try:
                resp = await client.post(
                    p["url"],
                    headers={"Authorization": f"Bearer {p['key']}",
                             "Content-Type": "application/json"},
                    json={"model": p["model"], "messages": messages,
                          "max_tokens": 800, "temperature": 0.3},
                )
                resp.raise_for_status()
                data = resp.json()
                return data["choices"][0]["message"]["content"], p["name"]
            except httpx.HTTPStatusError as e:
                last_error = f"{p['name']}: HTTP {e.response.status_code}"
                continue
            except Exception as e:
                last_error = f"{p['name']}: {e}"
                continue

    return (f"Lo siento, no puedo conectarme al servicio de IA en este momento "
            f"({last_error}). Intenta de nuevo en unos segundos.", None)


async def chat(user_message: str, conversation_history: list, app_state: dict) -> tuple[str, list, str | None]:
    """Función principal del chatbot. Devuelve (respuesta, historial, proveedor)."""
    contexto = get_context_for_query(user_message, app_state)
    system_prompt = build_system_prompt(contexto)

    messages = [{"role": "system", "content": system_prompt}]
    messages += conversation_history[-6:]
    messages.append({"role": "user", "content": user_message})

    respuesta, proveedor = await call_llm(messages)

    conversation_history.append({"role": "user", "content": user_message})
    conversation_history.append({"role": "assistant", "content": respuesta})
    return respuesta, conversation_history, proveedor

File: src/server/chatbot.ts
// Analytical chatbot: Groq primary -> Gemini fallback, lightweight RAG.
// Stateless. No API keys -> friendly message (never throws). Server-only.
import "server-only";

import { getEda, getPedidoScore, getSustitutos } from "~/server/data";

const GROQ_URL = "https://api.groq.com/openai/v1/chat/completions";
const GEMINI_URL = "https://generativelanguage.googleapis.com/v1beta/openai/chat/completions";
const GROQ_MODEL = "llama-3.3-70b-versatile";
const GEMINI_MODEL = "gemini-2.0-flash";

export type ChatMsg = { role: "system" | "user" | "assistant"; content: string };

async function buildContext(query: string) {
  const q = query.toLowerCase();
  const eda = await getEda();

  // Siempre: lo más útil para un tendero + resumen del negocio (es barato).
  const ctx: Record<string, unknown> = {
    resumen_general: eda.metricas_globales,
    productos_con_mas_cambios: eda.top_skus_por_tasa.slice(0, 12),
    cambios_mas_frecuentes: eda.top_pares.slice(0, 12),
  };

  if (/bodega|zona|centro|riesgo|problema|rojo|amarillo|segur|cedis/.test(q))
    ctx.bodegas_con_mas_cambios = eda.top_cedis.slice(0, 10);

  if (/país|pais|méxico|mexico|ecuador|perú|peru|argentina|colombia|región|region/.test(q))
    ctx.por_pais = eda.por_pais;

  if (/vendido|popular|volumen|más vende|mas vende|demanda|top/.test(q))
    ctx.productos_mas_vendidos = eda.top_skus.slice(0, 12);

  if (/estado|entrega|entregad|cancelad|pendiente|status|complet/.test(q))
    ctx.estados_de_pedidos = eda.status_pedidos;

  if (/reciente|tendencia|último|ultimo|recientemente|ahora/.test(q))
    ctx.recencia = eda.recencia;

  // Productos mencionados por nombre → incluye sus reemplazos más probables.
  const candidatos = new Set<string>();
  for (const p of eda.top_skus_por_tasa) candidatos.add(p.sku);
  for (const p of eda.top_pares) { candidatos.add(p.origen); candidatos.add(p.destino); }
  const mencionados = [...candidatos].filter((nombre) => {
    const primeras = nombre.toLowerCase().split(/[\s,]+/).slice(0, 2).join(" ");
    return primeras.length > 3 && q.includes(primeras);
  }).slice(0, 3);
  if (mencionados.length) {
    ctx.reemplazos_de_productos_mencionados = await Promise.all(
      mencionados.map(async (sku) => ({ producto: sku, reemplazos: await getSustitutos(sku, 3) })),
    );
  }

  // If query contains an order ID, fetch its score
  const m = query.replace(/[.,]/g, "").match(/\d{10,}/);
  if (m) {
    const ps = await getPedidoScore(m[0]!);
    if (ps) ctx.pedido_consultado = { id_pedido: m[0], ...ps };
  }
  return ctx;
}

function systemPrompt(ctx: Record<string, unknown>): string {
  return `Eres el asistente de Order Rescue, una app que ayuda a dueños de tienda, restaurantes y negocios a saber si sus pedidos de bebidas y productos Arca Continental llegarán completos.

Habla de forma amigable, simple y en español. Evita términos técnicos — usa "bodega" en vez de CEDIS, "producto" en vez de SKU, "cambio" o "reemplazo" en vez de "sustitución". El usuario es un tendero o dueño de negocio, no un técnico.

Puedes ayudar con:
- ¿Mi producto va a llegar o lo van a cambiar? (si dan un número de pedido)
- ¿Por qué producto me lo reemplazarían? ¿Cuál es el mejor sustituto?
- ¿Qué productos tienen más probabilidad de cambio? ¿Cuáles son los más vendidos?
- ¿Qué bodegas o países tienen más cambios?
- Comparar dos productos por riesgo de cambio.
- Recomendar qué conviene tener en stock para no quedarse corto.
- Dar un resumen ejecutivo del negocio (volumen, % de cambios, tendencias).

Cuando el riesgo de cambio es alto, tranquiliza al cliente y dile qué producto alternativo suele llegar en su lugar. Cuando todo está bien, confírmalo con seguridad. Si te piden una recomendación, sé concreto y accionable.

DATOS REALES DEL SISTEMA (úsalos para responder con precisión; no inventes cifras):
${JSON.stringify(ctx, null, 2)}

Reglas de formato y estilo:
- Nunca menciones términos como "score", "S1/S2/S3", "tasa de afectación" ni "CEDIS" — usa lenguaje cotidiano ("bodega", "producto", "cambio").
- Usa **negritas** en los números o nombres clave, y listas con viñetas cuando ayuden a leer.
- Si no tienes datos suficientes, dilo con amabilidad y sugiere usar el simulador de pedidos o dar el número de pedido.
- Respuestas cortas y directas (3-5 oraciones), salvo que pidan un análisis o resumen más amplio.`;
}

async function callLLM(messages: ChatMsg[]): Promise<{ text: string; provider: string | null }> {
  const providers = [
    { name: "Groq · llama-3.3-70b", url: GROQ_URL, model: GROQ_MODEL, key: (process.env.GROQ_API_KEY ?? "").trim() },
    { name: "Gemini · 2.0-flash", url: GEMINI_URL, model: GEMINI_MODEL, key: (process.env.GEMINI_API_KEY ?? "").trim() },
  ];

  if (!providers.some((p) => p.key)) {
    return {
      text: "⚠️ El asistente de IA no está configurado todavía. Agrega tu GROQ_API_KEY o GEMINI_API_KEY al archivo .env y reinicia el servidor para activarlo. Mientras tanto, puedes usar los tableros del dashboard.",
      provider: null,
    };
  }

  let lastError = "";
  for (const p of providers) {
    if (!p.key) continue;
    try {
      const ctrl = new AbortController();
      const timer = setTimeout(() => ctrl.abort(), 30000);
      const resp = await fetch(p.url, {
        method: "POST",
        headers: { Authorization: `Bearer ${p.key}`, "Content-Type": "application/json" },
        body: JSON.stringify({ model: p.model, messages, max_tokens: 800, temperature: 0.3 }),
        signal: ctrl.signal,
      });
      clearTimeout(timer);
      if (!resp.ok) { lastError = `${p.name}: HTTP ${resp.status}`; continue; }
      const data = (await resp.json()) as { choices?: { message?: { content?: string } }[] };
      const text = data.choices?.[0]?.message?.content;
      if (text) return { text, provider: p.name };
      lastError = `${p.name}: empty response`;
    } catch (e) {
      lastError = `${p.name}: ${e instanceof Error ? e.message : String(e)}`;
    }
  }
  return { text: `Lo siento, no puedo conectarme al servicio de IA en este momento (${lastError}). Intenta de nuevo en unos segundos.`, provider: null };
}

export async function chat(userMessage: string, history: ChatMsg[]) {
  const ctx = await buildContext(userMessage);
  const messages: ChatMsg[] = [
    { role: "system", content: systemPrompt(ctx) },
    ...history.slice(-6),
    { role: "user", content: userMessage },
  ];
  const { text, provider } = await callLLM(messages);
  const newHistory: ChatMsg[] = [
    ...history,
    { role: "user", content: userMessage },
    { role: "assistant", content: text },
  ];
  return { response: text, history: newHistory, provider };
}
